[scottg reviewing] Replace fetch data with preload terminology

nguyenhuy opened the pull request:
- Deprecate `-fetchData` and `-clearFetchedData` in favor of `-didEnterPreloadState` and `-didExitPreloadState`.
- `-didEnterPreloadState` and `-didExitPreloadState` call the deprecated `-fetchData` and `-clearFetchedData` methods if they are overriden.
- Move `-setNeedsPreload`, `-recursivelyPreload` and `-recursivelyClearPreloadedData` to `ASDisplayNode+FrameworkPrivate.h`. Those methods now call `-didEnterPreloadState` and `-didExitPreloadState`.
- Update internal implementation, comments and tests

Resolves #2611

nguyenhuy commented on AsyncDisplayKit/ASCollectionView.mm:
```diff
@@ -1675,18 +1675,6 @@ - (void)nodeDidRelayout:(ASCellNode *)node sizeChanged:(BOOL)sizeChanged
-{
-  [_rangeController clearFetchedData];
-}
-
```
These methods are not declared in ASCollectionView.h

appleguy commented on AsyncDisplayKit/ASDisplayNode+Subclasses.h:
```diff
@@ -324,21 +324,21 @@ NS_ASSUME_NONNULL_BEGIN
+ * selectively clear preloaded data.
  */
-- (void)clearFetchedData ASDISPLAYNODE_REQUIRES_SUPER;
+- (void)clearPreloadedData ASDISPLAYNODE_REQUIRES_SUPER;
```
Should we get rid of this method?  Is didExitPreloadState enough?  I think it probably is.

appleguy commented on AsyncDisplayKit/ASDisplayNode+Subclasses.h:
```diff
@@ -324,21 +324,21 @@ NS_ASSUME_NONNULL_BEGIN
  * The data may be remote and accessed via the network, but could also be a local database query.
  */
-- (void)fetchData ASDISPLAYNODE_REQUIRES_SUPER;
+- (void)preload ASDISPLAYNODE_REQUIRES_SUPER;
```
This one is trickier than I thought -- because of setNeedsDataFetch.  I think we should probably eliminate this method for now, and just call didEnterPreloadState again (even if imbalanced) when setNeedsDataFetch is called.  Let's deprecate setNeedsDataFetch and circle back to a cleaner API for "re-fetching" when we have more time (and we'll add that better API before finalizing removal of the deprecated method).

appleguy commented:
This is very close, thanks Huy!  Yes, having setNeedsPreload be in +FrameworkPrivate seems good for now.

Do you think we can avoid exposing -preload and -clearPreloadedData as public overrides?  Can we drive the didEnter and didExit methods instead, although this is a bit less conceptually clean, to give us more flexibility to improve the API later?

nguyenhuy commented:
@appleguy I pushed a new commit that addresses your comments. Yes, we can drive `-didEnterPreloadState` and `-didExitPreloadState` instead. It results in a cleaner API surface. We will need to revisit `-setNeedsPreload`, `-recursivelyPreload` and `-recursivelyClearPreloadedData` after 2.0, although they are "framework private" now.

nguyenhuy commented:
@appleguy Btw, looks like pinterest-jenkins is taking holidays. Machines don't have holidays!
